crates/holochain_websocket/src/websocket.rs

```diff
@@ -0,0 +1,763 @@
+use std::collections::VecDeque;
+use std::convert::TryInto;
+use std::sync::Arc;
+
+use futures::FutureExt;
+use futures::SinkExt;
+use futures::StreamExt;
+use holochain_serialized_bytes::prelude::*;
+use stream_cancel::Trigger;
+use stream_cancel::Valve;
+use stream_cancel::Valved;
+use tracing::instrument;
+use tracing::Instrument;
+
+use ghost_actor::*;
+use tungstenite::protocol::frame::coding::CloseCode;
+use tungstenite::protocol::CloseFrame;
+
+use crate::util::addr_to_url;
+use crate::util::ToFromSocket;
+use crate::util::CLOSE_TIMEOUT;
+use crate::CancelResponse;
+use crate::IncomingMessage;
+use crate::OutgoingMessage;
+use crate::RegisterResponse;
+use crate::Respond;
+use crate::WebsocketConfig;
+use crate::WebsocketError;
+use crate::WebsocketReceiver;
+use crate::WebsocketResult;
+use crate::WebsocketSender;
+use crate::WireMessage;
+
+type GhostResult<T> = std::result::Result<T, GhostError>;
+
+#[derive(Debug, Clone)]
+/// Actor that tracks responses.
+pub struct Websocket(GhostActor<WebsocketInner>);
+
+#[derive(Debug)]
+struct ResponseTracker {
+    /// List of responses where Some is an registered response
+    /// while None is a free slot.
+    responses: Vec<Option<RegisterResponse>>,
+    /// Queue of free slots to avoid ever growing memory.
+    free_indices: VecDeque<usize>,
```

Comment:
Consider a slow remote. We timeout a response / re-use an id. Now that previous response comes in, and is matched with the wrong request.

HashMaps are pretty efficient these days, I'd recommend using an always incrementing `u64` for message id, and storing the responses in a HashMap.

Reply:
Actually - I don't see any code to prune stale responses. We'll probably need to add that too.